bundles/org.dataflowanalysis.analysis.converter/src/org/dataflowanalysis/analysis/converter/PalladioConverter.java

```diff
@@ -0,0 +1,184 @@
+        if (source == null || dest == null) {
+            return;
+        }
+        List<DataFlowVariable> FlowVariables = pcmASE.getAllDataFlowVariables();
+        for (DataFlowVariable flowVariable : FlowVariables) {
+            String flowName = flowVariable.variableName();
+
+            Optional<Flow> optFlow = dataFlowDiagram.getFlows().stream().filter(f -> f.getSourceNode().equals(source))
+                    .filter(f -> f.getDestinationNode().equals(dest)).filter(f -> f.getEntityName().equals(flowName)).findFirst();
+
+            if (optFlow.isPresent()) {
+                return;
+            }
+
+            Flow newFlow = dataflowdiagramFactory.eINSTANCE.createFlow();
+            newFlow.setSourceNode(source);
+            newFlow.setDestinationNode(dest);
+            newFlow.setEntityName(flowName);
+
+            // Palladio Assumption: Each flows between two nodes with the same parameters/that are called the same use the same pin
+            Pin sourceOutPin = findOrCreateOutputPin(source, flowName);
+            Pin destInPin = findOrCreateInputPin(dest, flowName);
+            newFlow.setSourcePin(sourceOutPin);
+            newFlow.setDestinationPin(destInPin);
+
+            ForwardingAssignment forwarding = datadictionaryFactory.eINSTANCE.createForwardingAssignment();
+            forwarding.setOutputPin(sourceOutPin);
+            source.getBehaviour().getAssignment().add(forwarding);
+
+            this.dataFlowDiagram.getFlows().add(newFlow);
+        }
+    }
+
+    // A pin is equivalent if the same parameters are passed
+    private Pin findOrCreateOutputPin(Node source, String parameters) {
+        Optional<Pin> optPin = source.getBehaviour().getOutPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+        if (optPin.isPresent()) {
+            return optPin.get();
+        }
+        Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+        pin.setEntityName(parameters);
+        source.getBehaviour().getOutPin().add(pin);
+        return pin;
+    }
+
+    // A pin is equivalent if the same parameters are passed
+    private Pin findOrCreateInputPin(Node dest, String parameters) {
+        Optional<Pin> optPin = dest.getBehaviour().getInPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+        if (optPin.isPresent()) {
+            return optPin.get();
+        }
+        Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+        pin.setEntityName(parameters);
+        dest.getBehaviour().getInPin().add(pin);
+        return pin;
+    }
+
+    private Node getOrCreateDFDNode(AbstractPCMActionSequenceElement<? extends Entity> pcmASE) {
+        Node dfdNode = dfdNodeMap.get(pcmASE.getElement());
+
+        if (dfdNode == null) {
+            dfdNode = createCorrespondingDFDNode(pcmASE);
+            dfdNodeMap.put(pcmASE.getElement(), dfdNode);
+        }
```

Comment:
Use getOrDefault and putIfAbsent gere

Reply:
I've encountered something extraordinary:

This is fine:
```
Node dfdNode = dfdNodeMap.get(pcmASE.getElement());

if (dfdNode == null) {
      dfdNode = createDFDNode(pcmASE);
}

addNodeCharacteristicsToNode(dfdNode, pcmASE.getAllNodeCharacteristics());

return dfdNode;
```
This fails one of my equals assertions:
```
Node dfdNode = dfdNodeMap.getOrDefault(pcmASE.getElement(),createDFDNode(pcmASE));

addNodeCharacteristicsToNode(dfdNode, pcmASE.getAllNodeCharacteristics());

return dfdNode;
```
Is there a difference between the two?

Reply:
getOrDefault computes regardless of whether the node is present in the map.

If I use a compute if absent with a ConcurrentHashMap i get a recursive update exception

I would opt for doing it safely like in the working example above